services/sso/src/main/java/software.amazon.awssdk.services.sso/auth/SsoProfileCredentialsProviderFactory.java

```diff
@@ -0,0 +1,130 @@
+
+    private static final String TOKEN_DIRECTORY = Paths.get(userHomeDirectory(), ".aws", "sso", "cache").toString();
+
+    /**
+     * Default method to create the {@link SsoProfileCredentialsProvider} with a {@link SsoAccessTokenProvider}
+     * object created with the start url from {@link Profile} or environment variables and the default token file directory.
+     */
+    public AwsCredentialsProvider create(Profile profile) {
+        return create(profile, new SsoAccessTokenProvider(
+            generateCachedTokenPath(profile.properties().get(ProfileProperty.SSO_START_URL), TOKEN_DIRECTORY)));
+    }
+
+    /**
+     * Alternative method to create the {@link SsoProfileCredentialsProvider} with a customized
+     * {@link SsoAccessTokenProvider}. This method is only used for testing.
+     */
+    @SdkTestInternalApi
+    public AwsCredentialsProvider create(Profile profile,
+                                         SsoAccessTokenProvider tokenProvider) {
+        return new SsoProfileCredentialsProvider(profile, tokenProvider);
+    }
+
+    /**
+     * A wrapper for a {@link SsoCredentialsProvider} that is returned by this factory when {@link #create(Profile)} or
+     * {@link #create(Profile, SsoAccessTokenProvider)} is invoked. This wrapper is important because it ensures the parent
+     * credentials provider is closed when the sso credentials provider is no longer needed.
+     */
+    private static final class SsoProfileCredentialsProvider implements AwsCredentialsProvider, SdkAutoCloseable {
+        private final SsoClient ssoClient;
+        private final SsoCredentialsProvider credentialsProvider;
+
+        private SsoProfileCredentialsProvider(Profile profile,
+                                              SsoAccessTokenProvider provider) {
+            String ssoAccountId = profile.properties().get(ProfileProperty.SSO_ACCOUNT_ID);
+            String ssoRoleName = profile.properties().get(ProfileProperty.SSO_ROLE_NAME);
+            String ssoRegion = profile.properties().get(ProfileProperty.SSO_REGION);
+
+            this.ssoClient = SsoClient.builder()
+                                      .credentialsProvider(AnonymousCredentialsProvider.create())
+                                      .region(Region.of(ssoRegion))
+                                      .build();
+
+            GetRoleCredentialsRequest request = GetRoleCredentialsRequest.builder()
+                                                                         .accountId(ssoAccountId)
+                                                                         .roleName(ssoRoleName)
+                                                                         .build();
+
+            GetRoleCredentialsRequestSupplier supplier = new GetRoleCredentialsRequestSupplier(request,
+                                                                                               provider);
```

Comment:
You could simplify this to

```
Supplier<GetRoleCredentialsRequest> supplier = () -> request.toBuilder().accessToken(tokenProvider.resolveAccessToken()).build();
```

(or even use the copy() method instead of toBuilder() and build() if you're feeling extra frisky).

Reply:
Sounds good, will do the change.